### PR TITLE
make use of body as keyword a legacy feature

### PR DESCRIPTION
**Pull request by WalterBright**
with a warning when `-wo` is used.

### CODE REVIEW

File: compiler/src/dmd/expression.d [hunk 1]
```diff
@@ -2494,13 +2494,13 @@ extern (C++) final class SuperExp : ThisExp
...
 
-    override final bool isLvalue()
+    override bool isLvalue()
     {
         // Class `super` should be an rvalue
         return false;
     }
 
-    override final Expression toLvalue(Scope* sc, Expression e)
+    override Expression toLvalue(Scope* sc, Expression e)
```
Comment (ibuclaw) [inline compiler/src/dmd/expression.d:2503]: This looks to be unrelated.

File: compiler/src/dmd/mars.d [hunk 2]
```diff
@@ -153,9 +153,10 @@ private int tryMain(size_t argc, const(char)** argv, ref Param params)
     if (parseCommandlineAndConfig(argc, argv, params, files))
         return EXIT_FAILURE;
 
-    global.compileEnv.previewIn = global.params.previewIn;
-    global.compileEnv.ddocOutput = global.params.ddoc.doOutput;
+    global.compileEnv.previewIn        = global.params.previewIn;
+    global.compileEnv.ddocOutput       = global.params.ddoc.doOutput;
     global.compileEnv.shortenedMethods = global.params.shortenedMethods;
+    global.compileEnv.obsolete         = global.params.obsolete;
```
Comment (ibuclaw) [inline compiler/src/dmd/mars.d:159]: At some point, can't keep on copying across more params.

File: compiler/src/dmd/lexer.d [hunk 3]
```diff
@@ -51,6 +51,7 @@ struct CompileEnv
     bool previewIn;          /// `in` means `[ref] scope const`, accepts rvalues
     bool ddocOutput;         /// collect embedded documentation comments
     bool shortenedMethods = true;   /// allow => in normal function declarations
+    bool obsolete;           /// warn on use of legacy code
```
Comment (ibuclaw) [inline compiler/src/dmd/lexer.d:54]: I assume there's a header update missing, otherwise how is gdc or ldc going to set this?

File: compiler/src/dmd/parse.d [hunk 4]
```diff
@@ -9515,14 +9515,9 @@ LagainStc:
...
-        {
-            // @@@DEPRECATED_2.117@@@
-            // https://github.com/dlang/DIPs/blob/1f5959abe482b1f9094f6484a7d0a3ade77fc2fc/DIPs/accepted/DIP1003.md
-            // Deprecated in 2.097 - Can be removed from 2.117
-            // The deprecation period is longer than usual as `body`
-            // was quite widely used.
-            deprecation("usage of the `body` keyword is deprecated. Use `do` instead.");
+        if (compileEnv.obsolete)
+        {
+            eSink.warning(token.loc, "usage of identifer `body` as a keyword is obsolete. Use `do` instead.");
```
Comment (ibuclaw) [inline compiler/src/dmd/parse.d:9520]: I'm currently swaying towards rethinking how we do warnings/errors.  Possibly generating an enum from all the field members of `Param`, then adding overloads of these functions that have said enum as the first argument.

Think.
```
eSink.warning(GlobalParams.obsolete, token.loc, "usage of identifier `body` as a keyword is obsolete...");
```
It'll then be on the warning routine to map the enum value `GlobalParams.obsolete` back to its field, and check whether to emit the diagnostic or not.

File: compiler/src/dmd/mars.d (hunk 2 above)
Comment (WalterBright) [reply compiler/src/dmd/mars.d:159]: The idea is to remove all globals from the lexer/parser, and to provide to the parser only the minimum.

File: compiler/src/dmd/expression.d (hunk 1 above)
Comment (WalterBright) [reply compiler/src/dmd/expression.d:2503]: It is, but it is a complaint from from Dstyle that was overlooked before.

File: compiler/src/dmd/lexer.d (hunk 3 above)
Comment (WalterBright) [reply compiler/src/dmd/lexer.d:54]: done

File: compiler/src/dmd/parse.d (hunk 4 above)
Comment (WalterBright) [reply compiler/src/dmd/parse.d:9520]: The idea was to have eSink be an abstract interface, and have multiple implementations of it. This enabled lexer/parse/cparse to be independent of the baroque and overly complex implementation in errors.d. This was successful, and even allowed "mock" implementations for unittests that avoided problems like error gagging.

File: compiler/src/dmd/expression.d (hunk 1 above)
Comment (WalterBright) [reply compiler/src/dmd/expression.d:2503]: https://github.com/dlang/dmd/pull/15395

**Review by dkorpel — APPROVED** [review]

**Comment by WalterBright** [comment]
The buildkite error occurs across multiple PRs.
